### PR TITLE
Code Quality: Added StorageBar and StorageRing Controls

### PR DESCRIPTION
**Pull request by mdtauk**
**Resolved / Related Issues**
This adds StorageBar and StorageRing controls to the Files.App.Controls project for future use, as well as a test page in the UITests app for testing behaviour.

image

image

image

I am no expert when it comes to coding let alone making Controls, so I would appreciate these be given a thorough review to ensure there are no edge cases I have not anticipated, as well as general changes to improve the code by those with the expertise and experience. I will call out @yair100 and @0x5bfa  who have been seeing my progress via Discord as I was working on them.  But also anyone with control and or mathematics experience who can ensure I am using the most optimal  calculations for the angles and size calculations. @marcelwgn has contributed to the Community Toolkit so will be able to make some comments on my properties and general code quality.

I also want to indicate I took great inspiration from the RadialGuage control in the Toolkit, as well as the blogposts and sample code from **Diederik Krols** and his blog posts.
- https://xamlbrewer.wordpress.com/2018/12/06/creating-bindable-path-controls-in-uwp/
- https://xamlbrewer.wordpress.com/2016/12/16/a-percentage-ring-control-for-uwp/
- https://xamlbrewer.wordpress.com/2017/03/22/a-radial-range-indicator-control-for-uwp/

As well as Co-Pilot for maths queries and interpolation code suggestions, here are other posts I found helpful

- https://www.reddit.com/r/askmath/comments/17iqnae/circle_packing/
- https://www.vbforums.com/showthread.php?283960-Packing-circles-Resolved
- https://www.researchgate.net/figure/Circle-Packing-solution-from-N-8-to-N-13-and-equivalent-Point-Packing-problem-for-N_fig1_354328815
- https://thecolorofcode.com/xaml/wpfxaml-eine-runde-fortschrittsanzeige/

### CODE REVIEW

**Comment by mdtauk** [comment]
A final round of cleaning up or adding comments to parts of the controls can be done soon, but I wanted to get the functional code in after working on it on and off outside of the repo and adding it to the repo over the past week or so

File: src/Files.App.Controls/Storage/StorageRing/StorageRing.cs [hunk 1]
```diff
@@ -0,0 +1,1025 @@
...
+		/// <param name="MinAngle"></param>
+		/// <param name="MaxAngle"></param>
+		/// <returns></returns>
+		public static bool IsFullCircle(double MinAngle, double MaxAngle)
+		{
+			// Calculate the absolute difference between angles
+			double angleDifference = Math.Abs(MaxAngle - MinAngle);
+
+			// Check if the angle difference is equal to 360 degrees
+			return angleDifference == 360;
```
Comment (marcelwgn) [inline src/Files.App.Controls/Storage/StorageRing/StorageRing.cs:1020]: Comparing floating point numbers is a complex matter that is heavily discussed in computer science and more often than not, this comparison will not work as you expect. If you really need to be accurate a safer way to compare would be: `return Math.Abs(angleDifference - 360) < Double.Epsilon);` Given we are dealing with pixels and rendering, checking if we are within `1` or `0.1` of 360 should be fine.

File: src/Files.App.Controls/Storage/Data/ThicknessCheck.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+// Copyright (c) 2024 Files Community
+// Licensed under the MIT License. See the LICENSE.
+
+namespace Files.App.Controls
+{
+	/// <summary>
+	/// Defines ThicknessCheck values for <see cref="StorageRing"/> and <see cref="StorageBar"/>.
+	/// </summary>
+	public enum ThicknessCheck
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/Storage/Data/ThicknessCheck.cs:9]: I check which of the two thickness values is the thickest.  Value or Track, so I can adjust the controls based on this.

File: src/Files.App.Controls/Storage/RingShape/RingShape.cs [hunk 3]
```diff
@@ -0,0 +1,613 @@
...
+			return result;
+		}
+
+		/// <summary>
+		/// 
+		/// </summary>
+		/// <param name="d">The DependencyObject calling the function</param>
+		/// <param name="angle">The Angle to validate</param>
+		/// <param name="isStart">Bool to check if we are validating Start or End angle</param>
+		private void ValidateAngle(DependencyObject d, double angle, bool isStart)
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/Storage/RingShape/RingShape.cs:520]: I check to make sure the angle is between the minimum and maximum, clamping the value outside of the range.

File: src/Files.App.Controls/Storage/RingShape/RingShape.cs [hunk 4]
```diff
@@ -0,0 +1,613 @@
...
+		}
+
+		// Updates
+
+		/// <summary>
+		/// Suspends path updates until EndUpdate is called
+		/// </summary>
+		public void BeginUpdate()
+		{
+			_isUpdating = true;
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/Storage/RingShape/RingShape.cs:156]: The code sample I looked at before making this primitive path control, had it, and so I thought it was good practice.

File: src/Files.App.Controls/Storage/StorageBar/StorageBar.cs [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+		ColumnDefinition? _trackColumn;  // Reference to the TrackBar Column
+		ColumnDefinition? _gapColumn;    // Reference to the Gap Column
+
+		Border? _valueBarBorder;         // Reference to the Value Bar Border
+		Border? _trackBarBorder;         // Reference to the Track Bar Border
+
+		BarShapes _barShape;             // Reference to the BarShape
+
+		double _gapWidth;                // Stores the Gap between Value and Track Bars
+		double _smallerHeight;           // Stores the smaller between Value and Track Bars
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/Storage/StorageBar/StorageBar.cs:38]: I think they are private by default

File: src/Files.App.Controls/Storage/StorageBar/StorageBar.cs [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+			// Calculate the interpolation factor (t) between 0 and 1
+			var t = (value - startValue) / (endValue - startValue);
+
+			double interpolatedOutput;
+
+			if (useEasing)
+			{
+				// Apply an easing function (e.g., quadratic ease-in-out)
+				//var easedT = EaseInOutFunction(t);
+				var easedT = EaseOutCubic(t);
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/Storage/StorageBar/StorageBar.cs:567]: I was testing the two easing functions, we can remove the commented out one

File: src/Files.App.Controls/Storage/RingShape/RingShape.cs [hunk 7]
```diff
@@ -0,0 +1,613 @@
...
+		/// <param name="number">The input number.</param>
+		/// <param name="divider">The divider (non-zero).</param>
+		/// <returns>The positive modulus result.</returns>
+		private static double CalculateModulus(double number, double divider)
+		{
+			// Calculate the modulus
+			var result = number % divider;
+
+			// Ensure the result is positive or zero
+			result = result < 0 ? result + divider : result;
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/Storage/RingShape/RingShape.cs:509]: @0x5bfa How do you suggest I change the method to ensure it works correctly?

File: src/Files.App.Controls/Storage/RingShape/RingShape.cs [hunk 8]
```diff
@@ -0,0 +1,613 @@
...
+			CalculateAndSetNormalizedAngles(ringShape, ringShape.MinAngle, ringShape.MaxAngle);
+
+			ValidateAngle(ringShape, ringShape.StartAngle, true);
+			ValidateAngle(ringShape, ringShape.EndAngle, false);
+		}
+
+		/// <summary>
+		/// Updates the RingShape path
+		/// </summary>
+		private void UpdatePath()
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/Storage/RingShape/RingShape.cs:195]: I am breaking this up into smaller functions and updated comments

File: src/Files.App.Controls/Storage/StorageRing/StorageRing.cs [hunk 9]
```diff
@@ -0,0 +1,1025 @@
...
+		/// <summary>
+		/// Sets the Container size to the smaller of control's Height and Width.
+		/// </summary>
+		private void SetContainerSize(double controlWidth, double controlHeight, Thickness padding)
+		{
+			double correctedWidth = controlWidth - (padding.Left + padding.Right);
+			double correctedHeight = controlHeight - (padding.Top + padding.Bottom);
+
+			double check = Math.Min(correctedWidth, correctedHeight);
+			double minSize = 8;
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/Storage/StorageRing/StorageRing.cs:57]: I will do that

File: src/Files.App.Controls/Storage/StorageRing/StorageRing.cs [hunk 10]
```diff
@@ -0,0 +1,1025 @@
...
+		/// Initializes an instance of <see cref="StorageRing"/> class.
+		/// </summary>
+		public StorageRing()
+		{
+			DefaultStyleKey = typeof(StorageRing);
+
+			SizeChanged -= StorageRing_SizeChanged;
+			Unloaded -= StorageRing_Unloaded;
+			IsEnabledChanged -= StorageRing_IsEnabledChanged;
+			Loaded -= StorageRing_Loaded;
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/Storage/StorageRing/StorageRing.cs:108]: I will remove the unregister event handlers

File: src/Files.App.Controls/Storage/StorageRing/StorageRing.cs (hunk 1 above)
Comment (mdtauk) [reply src/Files.App.Controls/Storage/StorageRing/StorageRing.cs:1020]: I will make this change

File: src/Files.App.Controls/Storage/StorageRing/StorageRing.cs [hunk 11]
```diff
@@ -0,0 +1,1025 @@
...
+		}
+
+		/// <summary>
+		/// Converts a value within a specified range to a percentage.
+		/// </summary>
+		/// <param name="value">The value to convert.</param>
+		/// <param name="minValue">The minimum value of the input range.</param>
+		/// <param name="maxValue">The maximum value of the input range.</param>
+		/// <returns>The percentage value (between 0 and 100).</returns>
+		private double DoubleToPercentage(double value, double minValue, double maxValue)
```
[the remark this answers is not in the document]
Comment (mdtauk) [reply src/Files.App.Controls/Storage/StorageRing/StorageRing.cs:913]: I will do this in a future commit. @0x5bfa If you can help me go through the various calculation methods, and separate them into their own code file, I would greatly appreciate it.

File: src/Files.App.Controls/Storage/RingShape/RingShape.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App.Controls/Storage/RingShape/RingShape.cs:509]: If you're ensuring it be positive or zero, can't it be 'result < 0 ? 0 : result'?

**Review by Lamparter — CHANGES_REQUESTED** [review]

File: tests/Files.App.UITests/Views/StorageControlsPage.xaml [hunk 12]
```diff
@@ -0,0 +1,561 @@
...
+										Value="{x:Bind barTestValue.Value, Mode=TwoWay}" />
+
+								</Grid>
+							</Grid>
+
+						</Grid>
+
+						<AppBarSeparator Grid.Column="1" />
+
+
```
Comment (Lamparter) [inline tests/Files.App.UITests/Views/StorageControlsPage.xaml:417]:
```suggestion
```

File: tests/Files.App.UITests/Views/StorageControlsPage.xaml [hunk 13]
```diff
@@ -0,0 +1,561 @@
...
+								<TextBlock
+									Grid.Row="0"
+									Grid.Column="0"
+									Margin="24"
+									Style="{StaticResource TextExampleCaptionStyle}"
+									Text="LeftToRight" />
+
+
+
```
Comment (Lamparter) [inline tests/Files.App.UITests/Views/StorageControlsPage.xaml:466]:
```suggestion
```

File: tests/Files.App.UITests/Views/StorageControlsPage.xaml [hunk 14]
```diff
@@ -0,0 +1,561 @@
...
+									Value="{x:Bind barTestValue.Value, Mode=TwoWay}" />
+
+								<TextBlock
+									Grid.Row="0"
+									Grid.Column="0"
+									Margin="24"
+									Style="{StaticResource TextExampleCaptionStyle}"
+									Text="LeftToRight" />
+
+
```
Comment (Lamparter) [inline tests/Files.App.UITests/Views/StorageControlsPage.xaml:465]:
```suggestion
```

File: tests/Files.App.UITests/Views/StorageControlsPage.xaml [hunk 15]
```diff
@@ -0,0 +1,561 @@
...
+									Margin="24"
+									Style="{StaticResource TextExampleCaptionStyle}"
+									Text="RightToLeft" />
+
+							</Grid>
+
+						</Grid>
+
+
+
```
Comment (Lamparter) [inline tests/Files.App.UITests/Views/StorageControlsPage.xaml:496]:
```suggestion
```

File: tests/Files.App.UITests/Views/StorageControlsPage.xaml [hunk 16]
```diff
@@ -0,0 +1,561 @@
...
+									Grid.Column="0"
+									Margin="24"
+									Style="{StaticResource TextExampleCaptionStyle}"
+									Text="RightToLeft" />
+
+							</Grid>
+
+						</Grid>
+
+
```
Comment (Lamparter) [inline tests/Files.App.UITests/Views/StorageControlsPage.xaml:495]:
```suggestion
```

File: tests/Files.App.UITests/Views/StorageControlsPage.xaml [hunk 17]
```diff
@@ -0,0 +1,561 @@
...
+
+					</Grid>
+
+
```
Comment (Lamparter) [inline tests/Files.App.UITests/Views/StorageControlsPage.xaml:499]:
```suggestion
```

**Review by Lamparter — CHANGES_REQUESTED** [review]
Lots of extra spaces 😅

**Review by Lamparter — CHANGES_REQUESTED** [review]
Lots of extra spaces 😅

**Comment by 0x5bfa** [comment]
@Lamparter Please remove or close your all suggestions.

I'm going to take further steps for refactoring and simplification.

File: src/Files.App.Controls/Storage/RingShape/RingShape.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App.Controls/Storage/RingShape/RingShape.cs]: I think now is good?

File: src/Files.App.Controls/Storage/RingShape/RingShape.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App.Controls/Storage/RingShape/RingShape.cs]: What kind of modulus is this?

File: src/Files.App.Controls/Storage/RingShape/RingShape.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App.Controls/Storage/RingShape/RingShape.cs]: > Do we really need this isUpdating logic?

I'm sure we don't. Even if for interlock, UI gets updated on STA apartment and methods that call this method are called synchronous.

> The code sample I looked at before making this primitive path control, had it, and so I thought it was good practice.

@mdtauk Can you share the link?

In the meantime, I deleted.

File: src/Files.App.Controls/Storage/StorageBar/StorageBar.cs [hunk 18]
```diff
@@ -0,0 +1,599 @@
...
+		/// <summary>
+		/// Initializes an instance of <see cref="StorageBar"/> class.
+		/// </summary>
+		public StorageBar()
+		{
+			DefaultStyleKey = typeof(StorageBar);
+
+			SizeChanged -= StorageBar_SizeChanged;
+			Unloaded -= StorageBar_Unloaded;
+			IsEnabledChanged -= StorageBar_IsEnabledChanged;
```
[the remark this answers is not in the document]
Comment (0x5bfa) [reply src/Files.App.Controls/Storage/StorageBar/StorageBar.cs]: Removed

**Pull request merged**